addons/binding/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/handler/SamsungTvHandler.java

```diff
@@ -344,47 +290,193 @@ private SamsungTvService findServiceInstance(String serviceName) {
+                    RemoteControllerService.SERVICE_NAME);
+            if (service == null) {
+                service = RemoteControllerService.createNonUpnpService(configuration.hostName, configuration.port);
+            }
+
+            startService(service);
+            if (service.checkConnection()) {
+                putOnline();
             } else {
-                logger.trace("One or more services are already registered, not checking for new ones");
+                putOffline();
+                stopService(service);
             }
         } catch (RuntimeException e) {
             logger.warn("Catching all exceptions because otherwise the thread would silently fail", e);
         }
     }
 
-    private void startService(SamsungTvService service) {
+    private synchronized void startService(SamsungTvService service) {
         if (service != null) {
             service.addEventListener(this);
             service.start();
             services.add(service);
         }
     }
 
-    private void stopService(SamsungTvService service) {
+    private synchronized void stopService(SamsungTvService service) {
         if (service != null) {
             service.stop();
             service.removeEventListener(this);
             services.remove(service);
         }
     }
 
-    private void stopServices() {
-        logger.debug("Shutdown all Samsung services");
-        for (SamsungTvService service : services) {
-            stopService(service);
+    @Override
+    public void thingDiscovered(DiscoveryService source, DiscoveryResult result) {
+
+        if (configuration.hostName.equals(result.getProperties().get(SamsungTvConfiguration.HOST_NAME))) {
+            logger.debug("thingDiscovered: {}, {}", result.getProperties().get(SamsungTvConfiguration.HOST_NAME),
+                    result);
+            /*
+             * SamsungTV discovery services creates thing UID from UPnP UDN.
+             * When thing is generated manually, thing UID may not match UPnP UDN, so store it for later use (e.g.
+             * thingRemoved).
+             */
+            upnpThingUID = result.getThingUID();
+            logger.debug("thingDiscovered, thingUID={}, discoveredUID={}", this.getThing().getUID(), upnpThingUID);
+            checkAndCreateServices();
         }
-        services.clear();
     }
+
+    @Override
+    public void thingRemoved(DiscoveryService source, ThingUID thingUID) {
+        if (thingUID.equals(upnpThingUID)) {
+            logger.debug("Thing Removed: {}", thingUID);
+            shutdown();
+            putOffline();
+        }
+    }
+
+    @Override
+    public @Nullable Collection<@NonNull ThingUID> removeOlderResults(DiscoveryService source, long timestamp,
+            @Nullable Collection<@NonNull ThingTypeUID> thingTypeUIDs, @Nullable ThingUID bridgeUID) {
+        return null;
+    }
+
+    /**
+     * Send single WOL (Wake On Lan) package on all interfaces
+     */
+    void sendWOLAllInterfaces() {
```

Comment:
It feels sensible to add the WOL logic to a separate class, maybe outside of the bundle and in ESH?

Reply:
Agree to refactor it outside this class. I will not take it upon me to touch other classes outside the bundle: I am not familiar enough with the whole openHAB project and this goes outside the time I am willing to spend on this.